README.md

# modern_types [![skeleton](https://img.shields.io/badge/837babb-skeleton?label=%F0%9F%92%80%20bswck/skeleton&labelColor=black&color=grey&link=https%3A//github.com/bswck/skeleton)](https://github.com/bswck/skeleton/tree/837babb)
[![Package version](https://img.shields.io/pypi/v/modern-types?label=PyPI)](https://pypi.org/project/modern-types/)
[![Supported Python versions](https://img.shields.io/pypi/pyversions/modern-types.svg?logo=python&label=Python)](https://pypi.org/project/modern-types/)

[![Tests](https://github.com/bswck/modern_types/actions/workflows/test.yml/badge.svg)](https://github.com/bswck/modern_types/actions/workflows/test.yml)
[![Coverage](https://coverage-badge.samuelcolvin.workers.dev/bswck/modern_types.svg)](https://coverage-badge.samuelcolvin.workers.dev/redirect/bswck/modern_types)
[![Poetry](https://img.shields.io/endpoint?url=https://python-poetry.org/badge/v0.json)](https://python-poetry.org/)
[![Ruff](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/astral-sh/ruff/main/assets/badge/v2.json)](https://github.com/astral-sh/ruff)
[![License](https://img.shields.io/github/license/bswck/modern_types.svg?label=License)](https://github.com/bswck/modern_types/blob/HEAD/LICENSE)
[![Pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)](https://github.com/pre-commit/pre-commit)

> [!Warning]
> [There is a better solution for this](https://github.com/alexmojaki/eval_type_backport#readme).

`__modern_types__` aims to provide [PEP 585](https://peps.python.org/pep-0585/) + [PEP 604](https://peps.python.org/pep-0604/) backward compatibility for Python <=3.10 deferred type evaluation.
Hence, the targeted Python versions are 3.8 and 3.9.

# What does it do?
Technically speaking, `__modern_types__` traverses ASTs of type hint expressions to transform copies
of the namespaces passed to the evaluation routine `ForwardRef._evaluate`.
The transformation prevents type errors in Python 3.8 and 3.9 when evaluating these type hints using future-version PEP 585 and PEP 604 syntaxes.
This might be very useful for [writing pydantic models](https://docs.pydantic.dev/2.5/concepts/models/) in Python <3.10 in a modern fashion, without having to import `typing`.

As a result, in Python 3.8 and Python 3.9, the following code

```py
from __future__ import annotations

import collections.abc
from collections import defaultdict
from pprint import pprint
from typing import get_type_hints

import __modern_types__  # without this line it won't work!

class Foo:
    a: dict[str, int]
    b: list[int]
    c: set[int]
    d: tuple[int, ...] | None
    e: frozenset[int]
    f: defaultdict[str, int]
    g: str | None
    h: str | int
    i: str | int | None
    j: str
    k: collections.abc.Mapping[str, int]
    l: collections.abc.Mapping[str, int] | None
    m: collections.abc.Mapping[str, int | None] | float | None


pprint(get_type_hints(Foo, globals(), locals()))
```
gives:
```py
{'a': typing.Dict[str, int],
 'b': typing.List[int],
 'c': typing.Set[int],
 'd': typing.Union[typing.Tuple[int, ...], NoneType],
 'e': typing.FrozenSet[int],
 'f': typing.DefaultDict[str, int],
 'g': typing.Union[str, NoneType],
 'h': typing.Union[str, int],
 'i': typing.Union[str, int, NoneType],
 'j': <class 'str'>,
 'k': typing.Mapping[str, int],
 'l': typing.Union[typing.Mapping[str, int], NoneType],
 'm': typing.Union[typing.Mapping[str, typing.Union[int, NoneType]], float, NoneType]}
```
instead of raising an error that `type` object isn't subscriptable (Python 3.8)
or that `GenericAlias` doesn't support the `|` operator (Python 3.9).

# Use case
Keep your codebase up-to-date by speeding up migration to modern types, even if you support Python versions >=3.8.

Stop using deprecated `typing.Dict`, `typing.List`, `typing.Set`, `typing.Tuple`, `typing.FrozenSet`, `typing.DefaultDict` and other `typing` type proxies explicitly!

Importing `__modern_types__` will make all `typing.ForwardRef`-dependent parts of your application, including pydantic models, work flawlessly with PEP 585 and PEP 604.

# Is `__modern_types__` safe to use in production?
Yes. It doesn't break any existing codebase. It only uses AST and overwrites `typing.ForwardRef._evaluate`.
`__modern_types__` does not interact with the caller's namespaces, does not mutate built-in classes and does not do any other dubious things
that could potentially produce weird, unexpected side effects.

# How to use?
> [!Warning]
> Remember that the library does not change the built-in scope at runtime!
>
> So `dict[str, int]` won't render at runtime, but `typing.Dict[str, int]` will.
>
> `__modern_types__` makes it possible to evaluate `dict[str, int]` only through the `typing.get_type_hints` function.
>
> You should remember putting `from __future__ import annotations` at the top of your modules everywhere you
> want to leverage `__modern_types__`.

Simply import `__modern_types__` in your code, and it will make `typing.ForwardRef` instances go through the
type hint expression AST to try to tweak the copy of the passed global/local namespace
to use `typing._GenericAlias` instances that support `[]` and `|` operators at runtime.

Example replacements taking place in the built-in scope:

|             Old type              |           New type           | Without `__modern_types__`, works on Python version... | With `__modern_types__`, works on Python version... |                Backports PEP                 |
| :-------------------------------: | :--------------------------: | :----------------------------------------------------: | :-------------------------------------------------: | :------------------------------------------: |
|          `dict[KT, VT]`           |    `typing.Dict[KT, VT]`     |                         >=3.9                          |                        >=3.8                        | [PEP 585](https://peps.python.org/pep-0585/) |
|             `list[T]`             |       `typing.List[T]`       |                         >=3.9                          |                        >=3.8                        | [PEP 585](https://peps.python.org/pep-0585/) |
|             `set[T]`              |       `typing.Set[T]`        |                         >=3.9                          |                        >=3.8                        | [PEP 585](https://peps.python.org/pep-0585/) |
|          `tuple[T, ...]`          |    `typing.Tuple[T, ...]`    |                         >=3.9                          |                        >=3.8                        | [PEP 585](https://peps.python.org/pep-0585/) |
|          `frozenset[T]`           |    `typing.FrozenSet[T]`     |                         >=3.9                          |                        >=3.8                        | [PEP 585](https://peps.python.org/pep-0585/) |
|             `X \| Y`              |     `typing.Union[X, Y]`     |                       **>=3.10**                       |                        >=3.8                        | [PEP 604](https://peps.python.org/pep-0604/) |

Additionally, `__modern_types__` also allows you to use `collections.abc` and `contextlib` generic classes.

> [!Note]
> Some optional replacements will automatically also be registered if possible,
> according to those listed in the [`__modern_types__._typeshed`](https://github.com/bswck/modern_types/tree/HEAD/__modern_types__/_typeshed.py) source code.

# Installation



You might simply install it with pip:

```shell
pip install modern-types
```

If you use [Poetry](https://python-poetry.org/), then run:

```shell
poetry add modern-types
```

## For contributors

<!--
This section was generated from bswck/skeleton@837babb.
Instead of changing this particular file, you might want to alter the template:
https://github.com/bswck/skeleton/tree/837babb/fragments/readme.md
-->

> [!Note]
> If you use Windows, it is highly recommended to complete the installation in the way presented below through [WSL2](https://learn.microsoft.com/en-us/windows/wsl/install).



1.  Fork the [modern_types repository](https://github.com/bswck/modern_types) on GitHub.

1.  [Install Poetry](https://python-poetry.org/docs/#installation).<br/>
    Poetry is an amazing tool for managing dependencies & virtual environments, building packages and publishing them.
    You might use [pipx](https://github.com/pypa/pipx#readme) to install it globally (recommended):

    ```shell
    pipx install poetry
    ```

    <sub>If you encounter any problems, refer to [the official documentation](https://python-poetry.org/docs/#installation) for the most up-to-date installation instructions.</sub>

    Be sure to have Python 3.8 installed—if you use [pyenv](https://github.com/pyenv/pyenv#readme), simply run:

    ```shell
    pyenv install 3.8
    ```

1.  Clone your fork locally and install dependencies.

    ```shell
    git clone https://github.com/your-username/modern_types path/to/modern_types
    cd path/to/modern_types
    poetry env use $(cat .python-version)
    poetry install
    ```

    Next up, simply activate the virtual environment and install pre-commit hooks:

    ```shell
    poetry shell
    pre-commit install --hook-type pre-commit --hook-type pre-push
    ```

For more information on how to contribute, check out [CONTRIBUTING.md](https://github.com/bswck/modern_types/blob/HEAD/CONTRIBUTING.md).<br/>
Always happy to accept contributions! ❤️


# Legal info
© Copyright by Bartosz Sławecki ([@bswck](https://github.com/bswck)).
<br />This software is licensed under the terms of [MIT License](https://github.com/bswck/modern_types/blob/HEAD/LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).<br/>
This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!-- insertion marker -->

## [v2.0.8](https://github.com/bswck/modern_types/tree/v2.0.8) (2023-12-23)


No significant changes.


## [v2.0.7](https://github.com/bswck/modern_types/tree/v2.0.7) (2023-12-23)


### Changed

- Exported `PEP604GenericAliasLink` in `__modern_types__/__init__.py`.
- Corrected the README.

### Fixed

- Python version information in the README.
- `register()` breaking behavior on Python 3.8.


## [v2.0.6](https://github.com/bswck/modern_types/tree/v2.0.6) (2023-12-23)


### Changed

- * Exported `PEP604GenericAliasLink` in `__modern_types__/__init__.py`.
  * Corrected the README.

### Fixed

- Python version information in the README.
- `register()` breaking behavior on Python 3.8.


## [v2.0.5](https://github.com/bswck/modern_types/tree/v2.0.5) (2023-12-23)


No significant changes.


## [v2.0.4](https://github.com/bswck/modern_types/tree/v2.0.4) (2023-12-15)


### Fixed

- Moved flattening activation from tests to the main codebase.


## [v2.0.3](https://github.com/bswck/modern_types/tree/v2.0.3) (2023-12-15)


### Fixed

- Handling corner cases, such as when an external module is involved in the AST expression or non-generic types need unions.


## [v2.0.2](https://github.com/bswck/modern_types/tree/v2.0.2) (2023-12-12)


### Changed

- Once a type replacement is registered with `register()`, it must be manually removed from the registry to be changed by `register()`.


## [v2.0.1](https://github.com/bswck/modern_types/tree/v2.0.1) (2023-12-11)


No significant changes.


## [v2.0.0](https://github.com/bswck/modern_types/tree/v2.0.0) (2023-12-11)


### Fixed

- Rewrote the lib to use AST for overwriting namespaces, stopped mutating developer-end namespaces.


## [v1.0.8](https://github.com/bswck/modern_types/tree/v1.0.8) (2023-12-11)


### Fixed

- `recursive_guard` was added in 3.9.


## [v1.0.7](https://github.com/bswck/modern_types/tree/v1.0.7) (2023-12-11)


### Fixed

- Now works with [pydantic](https://pydantic.dev) which doesn't use `typing.get_type_hints`, but its own version that calls `typing._eval_type`.


## [v1.0.6](https://github.com/bswck/modern_types/tree/v1.0.6) (2023-12-11)


### Fixed

- `collections.defaultdict[int, str] | None` was impossible, because it was a 3.9 `types.GenericAlias` without `X | Y` union type syntax support.


## [v1.0.5](https://github.com/bswck/modern_types/tree/v1.0.5) (2023-12-11)


### Changed

- Significant README and docstring improvements.


## [v1.0.4](https://github.com/bswck/modern_types/tree/v1.0.4) (2023-12-11)


### Fixed

- Documentation.


## [v1.0.3](https://github.com/bswck/modern_types/tree/v1.0.3) (2023-12-11)


### Changed

- Enriched the README.


## [v1.0.2](https://github.com/bswck/modern_types/tree/v1.0.2) (2023-12-11)


### Added

- Some more description in the README.


## [v1.0.1](https://github.com/bswck/modern_types/tree/v1.0.1) (2023-12-11)


### Fixed

- `scripts/generate_auto.py` was mistakenly generating `__modern_types__/auto.py` instead of `__modern_types__/_auto.py`.


## [v1.0.0](https://github.com/bswck/modern_types/tree/v1.0.0) (2023-12-11)


### Added

- Full tests and implementation. ([#43283077](https://github.com/bswck/modern_types/issues/43283077))


__modern_types__/__init__.py
"""
`__modern_types__`.

`__modern_types__` aims to provide [PEP 585](https://peps.python.org/pep-0585/)
+ [PEP 604](https://peps.python.org/pep-0604/) backports for Python <=3.10
deferred type evaluation backward compatibility.

(C) 2023-present Bartosz Sławecki (bswck)
"""
from __future__ import annotations

import ast
import builtins
import sys
import typing
from functools import wraps
from types import SimpleNamespace

from __modern_types__._registry import (
    PEP604GenericAliasLink,
    PEP604Link,
    PEP604Proxy,
    create_dest_alias,
    global_registry,
    register,
)

__all__ = (
    "PEP604GenericAliasLink",
    "TypeHintVisitor",
    "register",
    "registry",
)


_PYTHON_VERSION = sys.version_info[:2]  # without PATCH version
_WARNING_3_10 = "You do not need to import __modern_types__ on Python >=3.10."

# ruff: noqa: N802


class _Bunch(SimpleNamespace):
    def __getitem__(self, key: str) -> object:
        return getattr(self, key)

    def __setitem__(self, key: str, value: object) -> None:
        setattr(self, key, value)


class TypeHintVisitor(ast.NodeTransformer):
    """
    Type hint visitor.

    Example.
    --------
    >>> import collections.abc
    >>> from __modern_types__ import TypeHintVisitor
    ...
    >>> global_ns = {"collections": collections}
    >>> local_ns = {}
    >>> visitor = TypeHintVisitor(global_ns, local_ns)
    >>> visitor.visit(ast.parse("a: collections.abc.Mapping[str, int]"))
    <_ast.Module object at 0x7f9b1c0b6d30>
    >>> local_ns
    {'a': typing.Mapping[str, int]}
    """

    registry: dict[object, object]

    def __init__(
        self,
        global_ns: dict[str, object],
        local_ns: dict[str, object],
        builtin_ns: dict[str, object] | None = None,
        registry: dict[object, object] | None = None,
    ) -> None:
        """Initialize TypeHintVisitor."""
        if builtin_ns is None:
            builtin_ns = vars(builtins)
        if registry is None:
            registry = global_registry
        self.registry = registry
        self.local_ns = local_ns
        self.global_ns = global_ns
        self.builtin_ns = builtin_ns

    @property
    def lookup(self) -> dict[str, object]:
        """Return a lookup dict for all namespaces."""
        return {**self.builtin_ns, **self.global_ns, **self.local_ns}

    def set_backport_for(  # noqa: C901, PLR0912
        self,
        path: list[str],
        obj: object | None = None,
        *,
        pep604: bool = False,
        pep604_name: str | None = None,
    ) -> None:
        """
        Set a backport object for an object at a given attribute path.

        Example.
        --------
        >>> import collections.abc
        >>> from __modern_types__ import TypeHintBackportSetter
        ...
        >>> backports = TypeHintBackportSetter({"collections": collections}, {}).
        >>> backports.set_backport_for(["collections", "abc", "Mapping"])
        >>> backports.global_ns["collections"].abc.Mapping
        typing.Mapping[~KT, ~VT_co]
        """
        if not path:
            return

        init_path = path[0]
        write_ns = None

        if init_path in self.local_ns:
            write_ns = self.local_ns
        elif init_path in self.global_ns or init_path in self.builtin_ns:
            write_ns = self.global_ns

        if write_ns is None:
            msg = f"namespace of {init_path!r} not found"
            raise LookupError(msg)

        current_ns: dict[str, object] | _Bunch = write_ns

        if len(path) > 1:
            *path_to, path_end = path[1:]

            try:
                ns = current_ns[init_path]
            except KeyError:
                return

            if not isinstance(ns, _Bunch):
                current_ns[init_path] = current_ns = _Bunch(**vars(ns))

            for name in path_to:
                next_ns = current_ns[name]
                if not isinstance(next_ns, _Bunch):
                    next_ns = current_ns[name] = _Bunch(**vars(next_ns))
                current_ns = next_ns
        else:
            path_end = init_path

        if obj is None:
            try:
                obj = current_ns[path_end]
            except KeyError:
                if len(path) == 1:
                    obj = self.builtin_ns[path_end]
                else:
                    raise

        if obj in self.registry:
            obj = self.registry[obj]

        if pep604 and not isinstance(obj, PEP604Link):
            obj = PEP604Proxy(
                obj,
                path_end,
                path_end := pep604_name or path_end,
            )

        current_ns[path_end] = obj

    def visit_Attribute(self, node: ast.Attribute) -> ast.Attribute:
        """Visit an attribute node. Note the behavior for children is different."""
        attr_path: list[str] = []
        for child in ast.walk(node):
            if isinstance(child, ast.Name):
                attr_path.insert(0, child.id)
            elif isinstance(child, ast.Attribute):
                attr_path.insert(0, child.attr)
            elif not isinstance(child, ast.Load):
                return node
        self.set_backport_for(attr_path)
        return node

    def visit_Name(self, node: ast.Name, *, pep604: bool = False) -> ast.Name:
        """Visit a name node. Can replace it with a backport."""
        if pep604:
            pep604_name = node.id
            while pep604_name in self.lookup:
                pep604_name += "__pep604"
            self.set_backport_for(
                [node.id],
                pep604=pep604,
                pep604_name=pep604_name,
            )
            if pep604_name in self.lookup:
                return ast.Name(id=pep604_name, ctx=node.ctx)
            return node
        self.set_backport_for([node.id])
        return node

    def pep604_visit(self, node: ast.AST) -> ast.AST:
        """Visit a node and replace it with a backport if needed."""
        if isinstance(node, ast.Name):
            return self.visit_Name(node, pep604=True)
        self.visit(node)
        return node

    def visit_BinOp(self, node: ast.BinOp) -> ast.BinOp:
        """Visit a binary operation node. Used for PEP 604 detection."""
        if isinstance(node.op, ast.BitOr):
            node.left = self.pep604_visit(node.left) or node.left  # type: ignore[assignment]
            self.visit(node.op)
            node.right = self.pep604_visit(node.right) or node.right  # type: ignore[assignment]
        else:
            self.generic_visit(node)
        return node


if _PYTHON_VERSION < (3, 10):
    _builtin_evaluate = typing.ForwardRef._evaluate  # noqa: SLF001

    @wraps(_builtin_evaluate)
    def _wrap_evaluate(
        self: typing.ForwardRef,
        globalns: dict[str, object] | None = None,
        localns: dict[str, object] | None = None,
        recursive_guard: frozenset[str] | None = None,
    ) -> typing.Any:
        """PEP 585 & PEP 604 backport."""
        # Copy the namespaces to avoid mutating them by TypeHintVisitor
        globalns = (globalns or {}).copy()
        localns = (localns or {}).copy()

        # The type hint to evaluate
        hint = ast.parse(self.__forward_arg__, mode="eval")

        # Perform the magic settings in globalns and localns
        TypeHintVisitor(globalns, localns).visit(hint)

        ast.fix_missing_locations(hint)

        self.__forward_code__ = compile(hint, filename="<type hint>", mode="eval")

        # Call the original _evaluate
        return create_dest_alias(
            _builtin_evaluate(
                self,
                globalns,
                localns,
                *() if _PYTHON_VERSION == (3, 8) else (recursive_guard or frozenset(),),
            ),
        )

    typing.ForwardRef._evaluate = typing.cast(typing.Any, _wrap_evaluate)  # type: ignore[attr-defined,method-assign,unused-ignore] #  noqa: SLF001

    from __modern_types__._registrars import register_typeshed_generics

    register_typeshed_generics()
else:
    import warnings

    warnings.warn(_WARNING_3_10, category=DeprecationWarning, stacklevel=-1)
